doc/man7/crypto.pod

```diff
@@ -12,41 +12,452 @@ See the individual manual pages for details.
+CMS, and they have also been used to implement many other third party products
+and protocols.
 
-C<libcrypto> consists of a number of sub-libraries that implement the
-individual algorithms.
+The functionality includes symmetric encryption, public key cryptography and key
+agreement, certificate handling, cryptographic hash functions, cryptographic
+pseudo-random number generators, message authentication codes (MACs), key
+derivation functions (KDFs), and various utilities.
 
-The functionality includes symmetric encryption, public key
-cryptography and key agreement, certificate handling, cryptographic
-hash functions, cryptographic pseudo-random number generator, and
-various utilities.
+=head2 Algorithms
 
-=head1 NOTES
+Cryptograpic primitives such as the SHA256 digest, or AES encryption are
+referred to in OpenSSL as "algorithms". An algorithm can have more than one
+implementation of it that is available for use. For example the RSA algorithm
+is available as a "default" implementation suitable for general use, and a
+"fips" implementation which has been validated to FIPS standards for situations
+where that is important. It is also possible that a third party could add
+additional implementations such as in a hardware security module (HSM).
 
-Some of the newer functions follow a naming convention using the numbers
-B<0> and B<1>. For example the functions:
+=head2 Operations
 
- int X509_CRL_add0_revoked(X509_CRL *crl, X509_REVOKED *rev);
- int X509_add1_trust_object(X509 *x, const ASN1_OBJECT *obj);
+Different algorithms can be grouped together by their purpose. For example there
+are algorithms for encryption, and different algorithms for digesting data.
+These different groups are known as "operations" in OpenSSL. Each operation
+has a different set of functions associated with it. For example to perform an
+encryption operation using AES (or any other encryption algorithm) you would use
+the encryption functions detailed on the L<EVP_EncryptInit(3)> page. Or to
+perform a digest operation using SHA256 then you would use the digesting
+functions on the L<EVP_DigestInit(3)> page.
 
-The B<0> version uses the supplied structure pointer directly
-in the parent and it will be freed up when the parent is freed.
-In the above example I<crl> would be freed but I<rev> would not.
+=head2 Providers
 
-The B<1> function uses a copy of the supplied structure pointer
-(or in some cases increases its link count) in the parent and
-so both (I<x> and I<obj> above) should be freed up.
+A provider in OpenSSL is a component that collects together algorithm
+implementations. In order to use an algorithm you must have at least one
+provider loaded that contains an implementation of it. OpenSSL comes with a
+number of providers and they may also be obtained from third parties. If you
+don't load a provider explicitly (either in program code or via config) then the
+OpenSSL built-in "default" provider will be automatically loaded.
```

Comment:
If you load any provider then the "default" provider would also need to be loaded if it is required. The legacy provider contains older algorithms and must be loaded if required. Algorithms such as MD2 and  RC4 are found in the legacy provider.

Reply:
I think this proposed text doesn't quite belong here - but I've split it up and added it to the more detailed text on the individual providers lower down this doc.